numpyro/primitives.py

```diff
@@ -224,38 +239,44 @@ def __init__(self, name, size, subsample_size=None, dim=None):
         self.subsample_size = size if subsample_size is None else subsample_size
         if dim is not None and dim >= 0:
             raise ValueError('dim arg must be negative.')
-        self.dim = dim
-        self._validate_and_set_dim()
+        self.dim, self._indices = self._subsample(self.name, self.size, self.subsample_size, dim)
         super(plate, self).__init__()
 
-    def _validate_and_set_dim(self):
+    # XXX: different from Pyro, this method returns dim and indices
+    @staticmethod
+    def _subsample(name, size, subsample_size, dim):
```

Comment:
I made this a staticmethod to be able to reuse it in `contrib.funsor`. The api is also more similar to Pyro.